### PR TITLE
Perform CIL=>LLVM translation using a symbolic execution engine.

### PR DESCRIPTION
**Pull request by mizho**

### CODE REVIEW

**Comment by lewurm** [comment]
hey @mizho, `CILSymbolicExecutor.cs` and `OpResultTypeLookup.cs` look pretty good! `LLVM/BitCodeEmitter.cs` is a bit of an overlap with `class Builder` from `BigStep.cs`, right?

`CILSymbolicExecutor` is still unused. How do you want to go forward from this?
* do you wanna write some basic tests?
* how do you wanna integrate it into the `BigStep` compiler?
* do you want to get it merged first?

File: mcs/class/Mono.Compiler/Mono.Compiler.BigStep/LLVM/BitCodeEmitter.cs [hunk 1]
```diff
@@ -0,0 +1,612 @@
...
+                {
+                    if (!BitCodeEmitter.s_uninitialized)
+                    {
+                        LLVM.LinkInMCJIT();
+                        LLVM.InitializeX86TargetMC();
+                        LLVM.InitializeX86Target();
+                        LLVM.InitializeX86TargetInfo();
+                        LLVM.InitializeX86AsmParser();
+                        LLVM.InitializeX86AsmPrinter();
+                        LLVM.InitializeMCJITCompilerOptions(s_options);
```
Comment (lewurm) [inline mcs/class/Mono.Compiler/Mono.Compiler.BigStep/LLVM/BitCodeEmitter.cs]: this only gives you x86 (32bit), we really want amd64 (64bit) code generated. so far, the generated code happens to be simple enough that it doesn't matter.

that's how you do it:
https://github.com/lambdageek/mono/pull/23/commits/ebe23b05edbbc33de98265caa0a38402146caa90

File: mcs/class/Mono.Compiler/Mono.Compiler.BigStep/LLVM/BitCodeEmitter.cs [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+        private static int s_moduleSeq;
+        private static bool s_uninitialized;
+
+        private LLVMModuleRef module;
+        private LLVMBuilderRef builder;
+        private LLVMValueRef function;
+
+        private LLVMValueRef[] argAddrs;
+        private LLVMValueRef[] localAddrs;
+        private Dictionary<string, LLVMValueRef> temps;
```
Comment (lambdageek) [inline mcs/class/Mono.Compiler/Mono.Compiler.BigStep/LLVM/BitCodeEmitter.cs]: Longer-term I kind of wish these stringly-keyed dictionaries were strongly typed.

File: mcs/class/Mono.Compiler/Mono.Compiler.BigStep/LLVM/BitCodeEmitter.cs [hunk 3]
```diff
@@ -0,0 +1,612 @@
...
+        private Dictionary<string, LLVMBasicBlockRef> bbs;
```
Comment (lambdageek) [inline mcs/class/Mono.Compiler/Mono.Compiler.BigStep/LLVM/BitCodeEmitter.cs]: also here the key shouldn't be a string, long term.

**Pull request merged**